homeassistant/components/restore_entity.py

```diff
@@ -0,0 +1,101 @@
+"""
+Support for restoring entity states on startup.
+
+Component that records all events and state changes. Allows other components
+to query this database.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/recorder/
+"""
+import logging
+from datetime import timedelta
+# from typing import Any, Union, Optional, List
+
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant
+from homeassistant.const import EVENT_HOMEASSISTANT_START
+from homeassistant.components import recorder
+from homeassistant.components.history import (
+    FILTER_SCHEMA, CONF_DOMAINS, CONF_EXCLUDE, CONF_ENTITIES, CONF_INCLUDE,
+    Filters, get_states)
+# import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.typing import ConfigType
+# import homeassistant.util.dt as dt_util
+
+DOMAIN = 'restore_entity'
+
+DEPENDENCES = ['recorder']
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: FILTER_SCHEMA,
+}, extra=vol.ALLOW_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Restore states."""
+    filters = Filters()
+    exclude = config[DOMAIN].get(CONF_EXCLUDE)
+    if exclude:
+        filters.excluded_entities = exclude[CONF_ENTITIES]
+        filters.excluded_domains = exclude[CONF_DOMAINS]
+    include = config[DOMAIN].get(CONF_INCLUDE)
+    if include:
+        filters.included_entities = include[CONF_ENTITIES]
+        filters.included_domains = include[CONF_DOMAINS]
+
+    # restore_entities = config.get(DOMAIN, {}).get(CONF_RESTORE, [])
+    def restore_states(event):
+        """Restore previous states from the DB."""
+
+        last_run = last_recorder_run()
+
+        last_end_time = last_run.end - timedelta(microseconds=1)
+        _LOGGER.debug("last run: start %s, end %s",
+                      last_run.start, last_end_time)
+
+        states = get_states(last_end_time, filters=filters)  # run=last_run,
+
+        # Not loading....
+        _LOGGER.debug('States loaded: %s', ', '.join(states))
+
+        no_state = []
+        for entity_id in states:
+            state = last_known_state(entity_id)
```

Comment:
Examples of [not mocking time of models)(https://travis-ci.org/home-assistant/home-assistant/jobs/181010724#L516) but cant find the DEBUG output from line 60 in Travis. I do see some SQLAlchemy exceptions, which wasnt so clear on my machine.... obviously still needs some work to add state entries before HASS start and ensure recorder starts up correctly during HASS start